README.md
ui7
===

[![npm](https://img.shields.io/npm/v/ui7?color=gray&label=%20&logo=npm)][npm]
[![deno.land](https://img.shields.io/github/v/tag/silverlyra/ui7?color=gray&label=%20&logo=deno&sort=semver)][deno]
[![GitHub Workflow Status (branch)](https://img.shields.io/github/workflow/status/silverlyra/ui7/CI/main?label=%20&logo=github)][build]
[![includes TypeScript types](https://img.shields.io/npm/types/ui7?color=333&label=%20&logo=typescript&logoColor=58baee)][typescript]
![node-current](https://img.shields.io/node/v/ui7?color=444&label=%20&logo=node.js)
[![MIT license](https://img.shields.io/npm/l/ui7?color=3ae)][license]

[npm]: https://www.npmjs.com/package/ui7
[deno]: https://deno.land/x/ui7
[build]: https://github.com/silverlyra/ui7/actions/workflows/ci.yml?query=branch%3Amain
[typescript]: https://deno.land/x/ui7/mod.ts
[license]: ./LICENSE

A small [UUIDv7][] generator, targeting the [latest draft][] update to [RFC 4122][]. v7 UUID’s are lexically sortable by their included timestamp.

[UUIDV7]: https://www.ietf.org/archive/id/draft-peabody-dispatch-new-uuid-format-04.html#name-uuid-version-7
[latest draft]: https://www.ietf.org/archive/id/draft-peabody-dispatch-new-uuid-format-04.html
[RFC 4122]: https://datatracker.ietf.org/doc/html/rfc4122

```js
import uuid, { timestamp } from "ui7";

const id = uuid();
// ==> "01836531-a895-7a2d-a70d-504ea62b40e2"

const when = new Date(timestamp(id));
// ==> 2022-09-22T12:34:56.789Z
```

## Installation

ui7 is available on [NPM][npm] and [deno.land][deno]:

#### npm

```sh
npm install --save ui7
```

Node.js versions 12.0 and up are supported, and _ui7_ includes an ESM build. It has no runtime dependencies.

#### deno

```ts
import uuid from "https://deno.land/x/ui7@v0.2.1/mod.ts";
```

## Usage


### Default generator

The default UUIDv7 generator is available as both a default export and a named `v7` function:

```js
import uuid from "ui7";

console.log(uuid());
// ==> "01836d65-3ec5-789f-b509-9ebd1d5ac2d3"
```

```js
import { v7 } from "ui7";

console.log(v7());
// ==> "01836d65-4a33-7aeb-aceb-8bfe42901787"
```

### Monotonic generators

Version 7 UUID’s include a millisecond-precision timestamp part and a random part. By default, _ui7_ will fully populate the random part with random bits. This means that ID’s generated during the same millisecond will sort in an arbitrary order:

```js
import uuid from "ui7";

const ids = [uuid(), uuid(), uuid()];
// [
//   "01836d99-1b31-75c7-bb83-aa52c651b085",
//   "01836d99-1b31-7b6a-8010-98fe59822b40",
//   "01836d99-1b31-71c4-bed3-6d3111c2de7d"
// ]

[...ids].sort().map(id => ids.indexOf(id))
// [ 2, 0, 1 ]
```

If you want UUID’s generated in the same process to sort in the order they were generated, you can instead create a _monotonic_ generator:

```js
import { generator } from "ui7";

// `generator` uses a monotonic entropy source by default.
const uuid = generator();

const ids = [uuid(), uuid(), uuid()];
// [
//   "01836da2-8a0d-7429-812d-f23f4d3d14c1",
//   "01836da2-8a0d-742a-89a3-0442ac613d11",
//   "01836da2-8a0d-742b-85e0-476f85ee36d9"
// ]

[...ids].sort().map(id => ids.indexOf(id))
// [ 0, 1, 2 ]
```

This uses the 12-bit `rand_a` field of UUIDv7 as a counter. For the first timestamp generated in a particular millisecond, the high bit is set to 0, so that there are always at least 2¹¹ (2,048) sequential counter values available before the field overflows (and wraps back to 0).

### Retrieving the timestamp

Use the `timestamp` function to read the timestamp field of a UUIDv7:

```js
import { timestamp } from "ui7";

const when = timestamp("01836db9-3d80-7df2-9c5e-d66442a3cf21");
// 1663993200000

const date = new Date(when);
// 2022-09-24T04:20:00.000Z
```

If the input to `timestamp` is not a UUIDv7, it throws a `ParseError` exception.

### Customizing generation

The default `uuid` function, the `generator` factory function, and the functions returned by `generator` all accept some options to customize their behavior.

#### Setting the time

You can pass a numeric timestamp, `Date` object, or custom clock function to a UUID generator or factory:

```js
import uuid from "ui7";

const id1 = uuid(0);
// "00000000-0000-7e2f-898b-a1f288a85661"

const id2 = uuid(new Date("2009-08-23T03:58:16.491Z"));
// "01234567-89ab-772e-9998-11f7395091b3"

const id3 = uuid(() => Date.now() * 2);
// "0306db58-42bc-722f-b2a5-4b13763cc462"

// Pass an object with a `time` property to combine with other options:
const id4 = uuid({ time: 1663993200000 });
// "01836db9-3d80-74a3-8b2a-8bf75bae40d5"
```

#### Controlling the format

You can omit dashes from the generated UUID’s, or use upper-case hex characters:

```js
import uuid from "ui7";

const id1 = uuid({ dashes: false });
// "01836db3b12d7f4e88c9fdc29019862e"

const id2 = uuid({ upper: true });
// "01836DB3-E2C7-7C84-8EB7-5CBE9205615B"

const id3 = uuid({ dashes: false, upper: true });
// "01836DB44DEF7836B69838F38A0A2A2A"
```

#### Using a different entropy source

By default, _ui7_ uses [`crypto.getRandomValues`][webcrypto] to populate the random part of the UUID (or [`crypto.randomFillSync`][fillsync] on Node.js versions that don't support Web-compatible crypto).

You can override this by providing an `entropy` function, which must return a [`Uint8Array`][u8a] with the requested `byteLength`:

```js
import uuid from "ui7";

const id1 = uuid({ entropy: (n) => new Uint8Array(n) });
// "01836db8-4831-7000-8000-000000000000"

const id2 = uuid({ entropy: (n) => new Uint8Array(n).map((_, i) => i) })
// "01836dbb-bd3d-7001-8203-040506070809"

// The entropy function receives the timestamp of the ID being generated:
const id4 = uuid({
  entropy(n, time) {
    const b = new Uint8Array(n);
    new DataView(b.buffer).setBigUint64(0, BigInt(time));
    return b;
  }
});
// "01836dbd-448c-7000-8183-6dbd448c0000"
```

[webcrypto]: https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues
[fillsync]: https://nodejs.org/api/crypto.html#cryptorandomfillsyncbuffer-offset-size
[u8a]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array

#### Creating a preset

In addition to providing options to the default generator, you can pass them to the `generator` factory function, which will return a generator that uses those options as its defaults:

```ts
import { generator } from "ui7";

const uuid = generator({
  time: () => Math.trunc(Date.now() / 1000),
  entropy: (n) => new Uint8Array(n).map(() => 0xFF),
  upper: true,
});

const id1 = uuid();
// "0000632F-3ACC-7FFF-BFFF-FFFFFFFFFFFF"

// Options can still be set for each individual ID:
const id2 = uuid({ dashes: false, time: 0 }):
// "0000000000007FFFBFFFFFFFFFFFFFFF"

```

### TypeScript

_ui7_ is written in TypeScript and ships with type definitions.

## Should I use this?

UUID’s are currently defined by [RFC 4122][] (2005), which specifies three UUID versions in common use: 1, 4, and 5. The “UUIDv7” identifiers generated by this package conform to [a candidate update][ietf tracker] to that RFC, and are not a generally-accepted standard.

UUIDv7 – a sortable UUID containing a UNIX timestamp – was first proposed in the second draft from April 2021. Originally, UUIDv7 held a variable-precision timestamp, but this was simplified to millisecond precision only in the fourth draft from March 2022. The latest draft made [minor revisions][dcl] in June 2022 and expires at the end of the year. It has not yet entered the formal process of advancing to a published RFC – UUIDv7 may never become standard.

There are advantages to the UUIDv7 format: they encode a timestamp, while still being collision-resistant with up to 74 bits of entropy per UUID. Their natural sortability by putting the timestamp field first is convenient for display and efficient for insertion into the [B-tree][] structures that underlie many databases. And they're still UUID's, and can be used directly in (e.g.) a Postgres `uuid` column without conversion.

But you might reasonably decide to say “no” or “not yet” to UUIDv7 until it is specified in an accepted RFC, or to use a different unique identifier format like [ULID][].

### Should I use this implementation?

_ui7_ is a compact, zero-dependency implementation of UUIDv7, using a cryptographic PRNG provided by its host environment. It is one of the simplest, but neither the fastest nor the slowest implementation; [a few others][others] can be found on NPM.

[ietf tracker]: https://datatracker.ietf.org/doc/draft-peabody-dispatch-new-uuid-format/
[dcl]: https://www.ietf.org/archive/id/draft-peabody-dispatch-new-uuid-format-04.html#section-3.1-2
[B-tree]: https://www.cs.cornell.edu/courses/cs3110/2012sp/recitations/rec25-B-trees/rec25.html
[ULID]: https://github.com/ulid/spec#readme
[others]: https://www.npmjs.com/search?q=uuidv7


CHANGELOG.md
# Changelog

## v0.2.1 (2022-09-24)

- Fixed the generation of monotonic UUID’s.

## v0.2.0 (2022-09-24)

- Added `bufferedRandom`, which calls `crypto.getRandomBytes` in blocks instead of on every UUID generation call, to improve performance significantly.
- Changed the signature of `EntropySource` to accept the number of desired random bytes.

## v0.1.0 (2022-09-23)

- Initial release
- Wrote the [readme](./README.md)


src/mod.test.ts
import test from 'ava';

import uuid, { generator, ParseError, pattern, timestamp } from './mod';

test('generates the RFC example UUID', (t) => {
  const time = 0x17f22e279b0;
  const entropy = () =>
    new Uint8Array([0x0c, 0xc3, 0x18, 0xc4, 0xdc, 0x0c, 0x0c, 0x07, 0x39, 0x8f]);

  t.is(uuid({ time, entropy }), '017f22e2-79b0-7cc3-98c4-dc0c0c07398f');
  t.is(uuid({ time, entropy, dashes: false }), '017f22e279b07cc398c4dc0c0c07398f');
  t.is(uuid({ time, entropy, upper: true }), '017F22E2-79B0-7CC3-98C4-DC0C0C07398F');
  t.is(uuid({ time, entropy, version: 8 }), '017f22e2-79b0-8cc3-98c4-dc0c0c07398f');
});

test('generates distinct UUIDs', (t) => {
  const a = uuid();
  const b = uuid();

  t.is(pattern.test(a), true);
  t.is(pattern.test(b), true);
  t.not(a, b);
});

test('generates UUIDs with the current timestamp', (t) => {
  const [now, id] = [Date.now(), uuid()];

  t.is(timestamp(id), now);
});

test('generates monotonic UUIDs', (t) => {
  const uuid = generator({ time: () => 0 });

  const ids = [...Array(2 ** 11)].map(uuid);
  t.is(pattern.test(ids[0]), true);

  const sorted = [...ids].sort();
  t.deepEqual(ids, sorted);
});

test('parses the timestamp field', (t) => {
  t.is(timestamp('017f22e2-79b0-7cc3-98c4-dc0c0c07398f'), 0x17f22e279b0);
  t.is(timestamp('017F22E2-79B0-7CC3-98C4-DC0C0C07398F'), 0x17f22e279b0);
  t.is(timestamp('017f22e279b07cc398c4dc0c0c07398f'), 0x17f22e279b0);
  t.is(timestamp('017F22E279B07CC398C4DC0C0C07398F'), 0x17f22e279b0);

  t.throws(() => timestamp('ea761762-3b05-11ed-a261-0242ac120002'), {
    instanceOf: ParseError,
  });
  t.throws(() => timestamp('0cddf0a0-d1b3-4688-8a11-c845855caa95'), {
    instanceOf: ParseError,
  });
  t.throws(() => timestamp('017f22e2-79b0-7'), {
    instanceOf: ParseError,
  });
});

test('matches the RFC example UUID', (t) => {
  t.is(pattern.test('017f22e2-79b0-7cc3-98c4-dc0c0c07398f'), true);
  t.is(pattern.test('017F22E2-79B0-7CC3-98C4-DC0C0C07398F'), true);
  t.is(pattern.test('017f22e279b07cc398c4dc0c0c07398f'), true);
  t.is(pattern.test('017F22E279B07CC398C4DC0C0C07398F'), true);

  // containing an invalid character
  t.is(pattern.test('017f22e2-79b0-7cc3-98c4-dc0c0c07398z'), false);
  t.is(pattern.test('017F22E2-79B0-7CC3-98C4-DC0C0C07398Z'), false);
  t.is(pattern.test('017f22e279b07cc398c4dc0c0c07398z'), false);
  t.is(pattern.test('017F22E279B07CC398C4DC0C0C07398Z'), false);

  // too long
  t.is(pattern.test('017f22e2-79b0-7cc3-98c4-dc0c0c07398fa'), false);
  t.is(pattern.test('017F22E2-79B0-7CC3-98C4-DC0C0C07398Fa'), false);
  t.is(pattern.test('017f22e279b07cc398c4dc0c0c07398fa'), false);
  t.is(pattern.test('017F22E279B07CC398C4DC0C0C07398Fa'), false);

  // too short
  t.is(pattern.test('017f22e2-79b0-7cc3-98c4-dc0c0c07398'), false);
  t.is(pattern.test('017F22E2-79B0-7CC3-98C4-DC0C0C07398'), false);
  t.is(pattern.test('017f22e279b07cc398c4dc0c0c07398'), false);
  t.is(pattern.test('017F22E279B07CC398C4DC0C0C07398'), false);
});


src/mod.ts
/**
 * Generate a UUIDv7.
 *
 * ## Options
 *
 * To generate a UUID at a specific time, you can pass a `number` timestamp or
 * `Date` object to this function.
 *
 * You can also further customize UUID generation by providing
 * {@link Options an object} with your preferences:
 *
 * - `time`:    Generate with this time instead of the current system time. You can
 *              provide a `number` millisecond-precision UNIX timestamp (as `Date.now`
 *              returns), a Date object, or a function returning a `number` timestamp.
 * - `dashes`:  `true` to include `-` characters in the generated UUID string;
 *              `false` to omit them. (default: `true`)
 * - `upper`:   Capitalize the A-F characters in the UUID. (default: `false`)
 * - `version`: The value of the UUID `version` field (default: `7`)
 * - `entropy`: A function to generate the random part of the UUID. Must return
 *              a `Uint8Array` containing 10 bytes.
 *              (default: uses `crypto.getRandomValues`)
 */
export const v7: Generator = (opt?: Clock | Time | Options | null): string => {
  const time = getTime(opt);
  let dashes = true;
  let upper = false;
  let version = 7 << 4;
  let rand: EntropySource = random;

  if (typeof opt === 'object' && opt !== null && !(opt instanceof Date)) {
    if (opt.dashes != null) dashes = opt.dashes;
    if (opt.version != null) version = (opt.version & 0x0f) << 4;
    if (opt.upper != null) upper = opt.upper;
    if (opt.entropy != null) rand = opt.entropy;
  }

  let timestamp = hex(time, 12);
  if (dashes) timestamp = timestamp.slice(0, 8) + '-' + timestamp.slice(8);

  const suffixBytes: string[] = Array(10);
  rand(10, time).forEach((b, i) => {
    if (i === 0) {
      b = version | (b & 0x0f);
    } else if (i === 2) {
      b = variant | (b & 0x3f);
    }
    suffixBytes[i] = hex(b);
  });

  const suffix = suffixBytes.join('');

  const id = dashes
    ? `${timestamp}-${suffix.slice(0, 4)}-${suffix.slice(4, 8)}-${suffix.slice(8)}`
    : timestamp + suffix;

  return upper ? id.toUpperCase() : id;
};

export default v7;

/**
 * Create a UUID generator. By default, uses a {@link monotonic} entropy source. Accepts the same options as the default generator.
 */
export const generator = (options?: GeneratorOptions): Generator => {
  options = {
    entropy: monotonic(),
    ...options,
  };

  return (opt?: Clock | Time | Options | null) => {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    if (opt == null) return v7(options!);

    if (opt instanceof Date || typeof opt === 'number' || typeof opt === 'function')
      return v7({ ...options, time: opt });

    return v7({ ...options, ...opt });
  };
};

/**
 * Return the timestamp portion of the UUID (a millisecond-precision UNIX
 * timestamp, as returned by `Date.now`).
 *
 * Throws a {@link ParseError} if no timestamp can be extracted.
 */
export const timestamp = (uuid: string): number => {
  const match = pattern.exec(uuid);
  if (match == null) throw new ParseError('Invalid v7 UUID; cannot determine timestamp');

  const ts = match[1].replace('-', '');
  return parseInt(ts, 16);
};

/**
 * A regular expression that recognizes UUID's generated by this package.
 *
 * Capture group `1` is the timestamp portion, and `2` is the random portion
 * (including the 2-bit constant UUID variant field).
 */
export const pattern =
  /^([0-9a-f]{8}-?[0-9a-f]{4})-?7([0-9a-f]{3}-?[0-9a-f]{4}-?[0-9a-f]{12})$/i;

/** The type of {@link v7 the UUID function} of this package. */
export interface Generator {
  /** Generate a UUIDv7 with default options. */
  (): string;

  /** Generate a UUIDv7 using the given time as its timestamp. */
  (time: Clock | Time | null | undefined): string;

  /** Generate a UUIDv7 using the given {@link Options options}. */
  (options: Options): string;
}

/**
 * A function that returns a millisecond-precision UNIX timestamp, like
 * `Date.now`.
 */
export type Clock = typeof Date['now'];

/** A `Date` object or millisecond-precision UNIX timestamp. */
export type Time = number | Date;

/** {@link v7 UUID generation} options. */
export interface Options extends FormatOptions {
  /**
   * Set the timestamp portion of the UUID to the given `Date`, UNIX timestamp
   * (as returned by `Date.now`), or the timestamp returned by the given
   * {@link Clock clock function}.
   */
  time?: Clock | Time;

  /**
   * Generate the random part of the UUID. The returned `Uint8Array` must be at
   * least 10 bytes long.
   */
  entropy?: EntropySource;
}

/** Configures a UUID {@link generator}. */
export interface GeneratorOptions extends FormatOptions {
  /** Use a different timestamp source than `Date.now`. */
  time?: Clock;

  /**
   * Generate the random part of the UUID. The returned `Uint8Array` must be at
   * least 10 bytes long.
   */
  entropy?: EntropySource;
}

/** Configures how a generated UUID is formatted. */
export interface FormatOptions {
  /**
   * `true` to include dashes in the UUID; `false` to omit them.
   * (default: `true`)
   */
  dashes?: boolean;

  /** Capitalize the A-F characters in the UUID. (default: `false`) */
  upper?: boolean;

  /** Set the version field of the UUID. (default: `7`) */
  version?: 7 | 8;
}

/**
 * A source for the `rand` fields of a UUID. To implement monotonic or other
 * counter-based `rand` fields, the UUID's `timestamp` is provided.
 *
 * Must return a `Uint8Array` of `size` bytes.
 */
export type EntropySource = (size: number, timestamp: number) => Uint8Array;

/**
 * A buffered `EntropySource`. To reduce overhead, `bufferedRandom` calls the
 * underlying `getRandomValues` in chunks of `blockSize`, returning the
 * already-generated random bytes for future reads until the block is exhausted.
 */
export const bufferedRandom = (blockSize = 200): EntropySource => {
  let bytes: Uint8Array | undefined;
  let buffer: ArrayBuffer;
  let pos: number;

  return (size: number) => {
    if (size > blockSize) {
      bytes = undefined;
      blockSize = size;
    }
    if (!bytes) {
      bytes = new Uint8Array(blockSize);
      buffer = bytes.buffer;

      crypto.getRandomValues(bytes);
      pos = 0;
    }

    let next = pos + size;
    if (next >= blockSize) {
      const leftover = blockSize - pos;
      bytes.copyWithin(0, pos);

      crypto.getRandomValues(new Uint8Array(buffer, leftover));
      pos = 0;
      next = size;
    }

    const result = new Uint8Array(buffer, pos, size);
    pos = next;

    return result;
  };
};

/**
 * Fill the UUID's `rand` fields with random bits, using
 * [`getRandomValues`][random]. UUID's produced with this randomness source are
 * not monotonic; ID's produced within the same millisecond will not necessarily
 * be sortable in the order they were produced.
 *
 * [random]: https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues
 */
export const random = bufferedRandom();

/**
 * Return an entropy source that allocates 12 bits as a monotonic counter at the
 * start of the UUIDv7 random fields.
 *
 * When a new timestamp is rolled over, the high bit of this field is set to
 * `0`. In the "unluckiest" case where the initial random counter value is
 * `0x7FF`, 2048 counter values are available within the same (millisecond)
 * timestamp value before exhaustion.
 *
 * **Note:** This package does not currently handle overflow of the counter
 * field; after `0xFFF`, the next value is `0x000`.
 */
export const monotonic = (entropy: EntropySource = random): EntropySource => {
  let bytes = new Uint8Array(10);
  let randomBytes = new Uint8Array(bytes.buffer, 2);

  let lastTimestamp = 0;
  let seq: number | undefined;

  return (size, timestamp) => {
    if (bytes.byteLength !== size) {
      bytes = new Uint8Array(size);
      randomBytes = new Uint8Array(bytes.buffer, 2);
    }

    if (timestamp > lastTimestamp) {
      bytes.set(entropy(10, timestamp));
      bytes[0] &= 0x07;

      lastTimestamp = timestamp;
      seq = undefined;
    } else {
      if (seq === undefined) seq = ((bytes[0] & 0x0f) << 8) | bytes[1];
      seq++; // TODO: handle overflow

      randomBytes.set(entropy(8, timestamp));
      bytes[0] = (seq >> 8) & 0xff;
      bytes[1] = seq & 0xff;
    }

    return bytes;
  };
};

/** The exception thrown by {@link timestamp} if UUID parsing fails. */
export class ParseError extends Error {
  public readonly name = 'ParseError';
}

const getTime = (time: Clock | Time | Options | null | undefined): number => {
  if (time == null) return Date.now();
  if (typeof time === 'number') return time;
  if (time instanceof Date) return +time;
  if (typeof time === 'function') return time();
  return getTime(time.time);
};

const hex = (n: number, width = 2) => n.toString(16).padStart(width, '0');

const variant = 2 << 6;
